### PR TITLE
[97] First spike at GA V4 measurement protocol

### LINKED ISSUE #97: Support for Google Analytics v4
It seems to be in alpha, but wouldn't it be easy to add support for GA4 ? 

https://developers.google.com/analytics/devguides/collection/protocol/ga4

(It is referenced from the v1 protocol https://developers.google.com/analytics/devguides/collection/protocol/v1/ page)

### PR DESCRIPTION
**Pull request by tpitale**
Resolves #97 

* Item struct
* add_payment_info Event started
* porting tracker to a new v4 tracker
* All events listed have been added with their fields
* Dynamic methods on tracker for each event
* Net::HTTP adapter support
* Validation support

**TODO:**

* [ ] Support for all adapters
* [ ] README updates
* [ ] Firebase support?
* [ ] Alpha release

### CODE REVIEW

File: lib/staccato.rb [hunk 1]
```diff
@@ -59,16 +59,18 @@ def self.as_url(hit, uri = nil)
...
+require_relative 'staccato/event'
+require_relative 'staccato/social'
+require_relative 'staccato/exception'
+require_relative 'staccato/timing'
+require_relative 'staccato/transaction'
+require_relative 'staccato/transaction_item'
+require_relative 'staccato/tracker'
+require_relative 'staccato/measurement'
+
+require_relative 'v4'
```
Comment (tyrauber) [inline lib/staccato.rb:76]: @tpitale Shouldn't this be `require_relative 'staccato/v4'`?

File: lib/staccato/V4/event.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    # this module is included into each model hit type
+    #   to share the common behavior required to hit
+    #   the Google Analytics /collect api endpoint
+    def self.included(event)
+      event.extend Forwardable
+
+      event.class_eval do
+        attr_accessor :tracker, :options
+
+        def_delegators :@options, *event::FIELDS.keys
```
Comment (tyrauber) [inline lib/staccato/V4/event.rb:12]: @tpitale, I think this should just be `event::fields`, because it is an array of symbols instead of an object.

**Comment by tyrauber** [comment]
@tpitale Thank you for this. We've been using Staccato to instrument our API analytics to GA (thank you).  With the coming deprecation, we are looking to migrate our google analytics tags to v4, including our API analytics.  I've got a branch with the above changes, and I will try to update our implementation.  

With this change, should we be tracking events in v4 with ```Staccato::V4.tracker```?  It looks like `pageview` and `event` are not implemented?  Is the desire to be backwards compatible? Or are you thinking about an API change? 

Happy to help move this forward if I can. Thanks again for this library.

**Comment by tpitale** [comment]
@tyrauber Yeah, definitely in draft state. I've just been kind of playing around with it, since the protocol is so different.

**Comment by tpitale** [comment]
@tyrauber This will be an entirely new API. `pageview` and `event` are not listed in the new event types by google: https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference/events

They seem to have changed to just a list of more-specific events like `login` and `add_to_cart`, etc. According to their docs:

> The purpose of measurement protocol is to augment existing events collected via gtag, GTM, or Firebase.

So it doesn't seem like server-to-server is intended as the only method of collection.

**Comment by linhdangduy** [comment]
Hi, thank you for the pull request.

Any updates/plan for support GA4 measurement protocol?
Look like it in draft for a while.

**Comment by Paradoxor** [comment]
Hi, thank you for the PR. Are you planing on continuing this feature?
Because the deadline where google is shutting down the old api is moving closer and than this gem wouldn't work at all anymore. Would be really sad to see this amazing gem to be not usable anymore. 

Cheers

**Comment by tpitale** [comment]
I think this is ready for someone to try it out.

It only works with Net::HTTP adapter right now, but you should be able to build your events and call `tracker.validate!`. 🤞🏼 

This _might_ work:

```
tracker = Staccato::V4::Tracker.new(<measurement_id>, <api_secret>)

tracker.add_payment_info({currency: "USD", value: "99.00", payment_type: "Mastercard", items: []})
tracker.validate!
# or tracker.track
```

**Comment by tannerhuynh** [comment]
@tpitale 

I tried your method above using the search event

```
tracker = Staccato::V4::Tracker.new(<measurement_id>, <api_secret>)

tracker.search({search_term: <value>})
tracker.validate!
```

and I got the following error

```
eval error: uninitialized constant Staccato::Adapter
V4/tracker.rb:68 in validate!
```

**Comment by tpitale** [comment]
Thank you @tannerhuynh I will try to take a look at those soon. Appreciate you testing it out.

**Comment by tannerhuynh** [comment]
@tpitale NP, it should be an easy fix. You have to parse the URI. Once I did that locally and bundled the gem, the GA4 server we have is now receiving the correct hit.

**Comment by tannerhuynh** [comment]
@tpitale Sorry another question... 

In the events array, the code seems to be sending the entire object <Stacatto::V4::Search> instead of the expected payload JSON payload

What I see...

```
events: [ { "<Stacatto::V4::Search>" }  ] 
```

What MP expects...

```
events: [{
      name: 'search',
      params: { search_term: <term> },
    }]
```

**Comment by tpitale** [comment]
@tannerhuynh I think I've pushed fixes for all the issues you identified. Thank you!

**Comment by tannerhuynh** [comment]
Thanks @tpitale. The data seems to be correct in GA4. The only thing that isn't working is the validation endpoint. I wonder if its still hitting the old UA endpoint? There also seems to be too many arguments given for the Staccato adapter.

```
wrong number of arguments (given 2, expected 0..1) (ArgumentError)
.../staccato-f3986222ab78/lib/staccato/adapter/validate.rb:4:in `initialize'
.../staccato-f3986222ab78/lib/staccato/V4/tracker.rb:68:in `new'
.../staccato-f3986222ab78/lib/staccato/V4/tracker.rb:68:in `validate!'
```

File: lib/staccato/v4.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      })
+    end
+  end
+end
+
+require_relative 'option_set'
+require_relative 'v4/event'
+
+# define tracker last so any events have dynamically defined methods
+require_relative 'v4/tracker'
```
Comment (tannerhuynh) [inline lib/staccato/v4.rb:58]: The `v` here should be capitalized for

require_relative 'v4/event'
require_relative 'v4/tracker'

File: lib/staccato.rb [hunk 4]
```diff
@@ -59,16 +59,18 @@ def self.as_url(hit, uri = nil)
...
+require_relative 'staccato/event'
+require_relative 'staccato/social'
+require_relative 'staccato/exception'
+require_relative 'staccato/timing'
+require_relative 'staccato/transaction'
+require_relative 'staccato/transaction_item'
+require_relative 'staccato/tracker'
+require_relative 'staccato/measurement'
+
+require_relative 'staccato/v4'
```
Comment (tannerhuynh) [inline lib/staccato.rb:76]: I think this needs a new line at the end.

File: lib/staccato/v4.rb (hunk 3 above)
Comment (tpitale) [reply lib/staccato/v4.rb:58]: Oh yeah 😮‍💨

**Comment by tannerhuynh** [comment]
@tpitale are there any plans to support custom events and other params like user_id?

**Comment by tpitale** [comment]
I haven't looked at the docs for any of that. But I don't see why not. Probably not a part of this PR.

**Comment by phlegx** [comment]
Hi @tpitale! When is a release planned?

**Comment by tpitale** [comment]
@phlegx no release planned. As you can probably read in the comments I'm working on it when I can.

File: lib/staccato/V4/event.rb [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+require_relative 'events/item'
+require_relative 'events/add_payment_info'
+require_relative 'events/add_shipping_info'
+require_relative 'events/add_to_cart'
+require_relative 'events/add_to_wishlist'
+require_relative 'events/begin_checkout'
+require_relative 'events/earn_virtual_currency'
+require_relative 'events/generate_lead'
+require_relative 'events/item'
```
Comment (phlegx) [inline lib/staccato/V4/event.rb:83]: Duplicate require events/item. Already included on line 75.

**Comment by tpitale** [comment]
@tannerhuynh I think I resolved the validation adapter issue, could you try again?

@phlegx Would you be willing to test this PR if you have access to V4? I don't use GA anymore, so I don't really have a way of testing.

**Comment by phlegx** [comment]
Hi @tpitale 

I use it with custom metric and dimension like the following code. Do this PR support custom data?

```ruby

# variables payload, dimensions and metrics

tracker = Staccato.tracker('UA-XXXXXXXXXXX', nil, ssl: true)
hit = Staccato::Pageview.new(tracker, payload)
dimensions.each_with_index do |v, i|
  hit.add_custom_dimension(i + 1, v)
end
metrics.each_with_index do |v, i|
  hit.add_custom_metric(i + 1, v)
end
hit.track!
```

**Comment by tpitale** [comment]
@phlegx Yes, this is planned as a followup to this. See #101 changes.

**Pull request merged**